modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,200 @@
+
+    auto reply = this->callback_(internal::deserializeRequest<RequestT>(query));
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_EMPTY });  // Update encoding.
+      auto buffer = serdes::serialize(reply);
+      DLOG(INFO) << "Replying with buffer of size: " << buffer.size();
+      DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
+      query.reply(this->topic_config_.name, std::move(buffer), options);
+    }
+  };
+
+  queryable_ = expectAsUniquePtr(
+      session_->zenoh_session.declare_queryable(topic_config_.name, { std::move(query), []() {} }));
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(Session& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::optional<std::chrono::milliseconds>& timeout /*= std::nullopt*/)
+    -> std::vector<ServiceResponse<ReplyT>> {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  LOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  std::vector<ServiceResponse<ReplyT>> reply_messages;
+
+  auto on_reply = [&reply_messages, &m, &topic_config](zenohc::Reply&& reply) {
+    internal::onReply(std::move(reply), reply_messages, m, topic_config.name);
+  };
+
+  auto on_done = [&m, &done_signal]() {
+    std::lock_guard lock(m);
+    done_signal.notify_all();
+  };
+
+  zenohc::GetOptions options{};
+  if (timeout.has_value()) {
+    options.timeout_ms = static_cast<uint64_t>(timeout->count());
+  }
+
+  if constexpr (std::is_same_v<RequestT, std::string>) {
+    options.set_value(zenohc::Value(request, Z_ENCODING_PREFIX_TEXT_PLAIN));
+  } else {
+    auto buffer = serdes::serialize(request);
+    buffer = internal::addChangingBytes(buffer);
+
+    DLOG(INFO) << "Sending buffer of size: " << buffer.size();
+    DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
+    auto value = zenohc::Value(std::move(buffer), Z_ENCODING_PREFIX_EMPTY);
+    options.set_value(value);
+  }
+
+  zenohc::ErrNo error_code{};
+  const auto success =
+      session.zenoh_session.get(topic_config.name.c_str(), "", { on_reply, on_done }, options, error_code);
+
+  std::unique_lock lock(m);
+  if (timeout.has_value()) {
+    if (done_signal.wait_for(lock, timeout.value()) == std::cv_status::timeout) {
+      LOG(WARNING) << fmt::format("Timeout while waiting for service reply of '{}'", topic_config.name);
+      return {};
+    }
+  } else {
+    done_signal.wait(lock);
+  }
+
+  if (!success) {
+    LOG(ERROR) << fmt::format("Error while calling service on '{}': '{}'", topic_config.name, error_code);
+    return {};
+  }
```

Comment:
success returns immediately so this should be check right after the call with `throwExceptionIf`. 
Here we should check `error_code`